// @appsignal/plugin-breadcrumbs-network/dist/esm/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
function networkBreadcrumbsPlugin(options) {
    var opts = __assign({ xhrEnabled: true, fetchEnabled: true }, options);
    opts.xhrEnabled = opts.xhrEnabled ? "XMLHttpRequest" in window : false;
    opts.fetchEnabled = opts.xhrEnabled ? "fetch" in window : false;
    return function () {
        var _this = this;
        var xhrPatch = function () {
            var appsignal = _this;
            var prevOpen = XMLHttpRequest.prototype.open;
            function open(method, url) {
                var metadata = { method: method, url: url };
                function onXhrLoad() {
                    appsignal.addBreadcrumb({
                        action: this.status >= 400
                            ? "Request failed with code " + this.status
                            : "Recieved a response with code " + this.status,
                        category: "XMLHttpRequest",
                        metadata: metadata
                    });
                }
                function onXhrError() {
                    appsignal.addBreadcrumb({
                        action: "Request failed",
                        category: "XMLHttpRequest",
                        metadata: metadata
                    });
                }
                this.addEventListener("load", onXhrLoad);
                this.addEventListener("error", onXhrError);
                prevOpen.apply(this, arguments);
            }
            XMLHttpRequest.prototype.open = open;
        };
        var fetchPatch = function () {
            var appsignal = _this;
            var originalFetch = window.fetch;
            function fetch(input, init) {
                var url = typeof input === "string" ? input : input.url;
                var method = (typeof input !== "string" && input.method) ||
                    (init && init.method) ||
                    "GET";
                var metadata = {
                    method: method,
                    url: url
                };
                return originalFetch
                    .apply(window, arguments)
                    .then(function (response) {
                    var statusCode = response.status;
                    appsignal.addBreadcrumb({
                        action: "Recieved a response with code " + statusCode,
                        category: "Fetch",
                        metadata: metadata
                    });
                    return response;
                })
                    .catch(function (error) {
                    appsignal.addBreadcrumb({
                        action: "Request failed",
                        category: "Fetch",
                        metadata: metadata
                    });
                    throw error;
                });
            }
            window.fetch = fetch;
        };
        if (opts.xhrEnabled)
            xhrPatch();
        if (opts.fetchEnabled)
            fetchPatch();
    };
}
export var plugin = networkBreadcrumbsPlugin;
