// @appsignal/react/dist/esm/ErrorBoundary.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import React from "react";
var ErrorBoundary = (function (_super) {
    __extends(ErrorBoundary, _super);
    function ErrorBoundary() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { error: undefined };
        return _this;
    }
    ErrorBoundary.getDerivedStateFromError = function (error) {
        return { error: error };
    };
    ErrorBoundary.prototype.componentDidCatch = function (error, info) {
        var _a = this.props, appsignal = _a.instance, action = _a.action;
        var name = error.name, message = error.message, stack = error.stack;
        var span = appsignal.createSpan();
        span
            .setAction(action !== "" ? action : undefined)
            .setError({
            name: name,
            message: message,
            stack: stack
        })
            .setTags({ framework: "React" });
        appsignal.send(span);
        if (!this.state.error)
            this.setState({ error: error });
    };
    ErrorBoundary.prototype.render = function () {
        if (this.state.error) {
            return this.props.fallback ? this.props.fallback(this.state.error) : null;
        }
        return this.props.children;
    };
    ErrorBoundary.defaultProps = {
        action: ""
    };
    return ErrorBoundary;
}(React.Component));
export { ErrorBoundary };
